src/utils/reporting_workflow.js

```diff
@@ -0,0 +1,154 @@
+};
+
+function initializeContext(planDefinition) {
+  // TODO -- see MEDMORPH-36
+  return { id: uuidv4(), planDefinition };
+}
+
+function getFunction(ig, code) {
+  // use a function here in case this needs to be more complex
+  return IG_REGISTRY[ig][code];
+}
+
+/**
+ * Find the target profile to be used when generating a bundle
+ * in support of the given PlanDefinition.
+ * The profile should be on the 'output' field of the 'create-report' action.
+ */
+function findProfile(planDefinition) {
+  // figure out the profile of the target bundle
+
+  // TODO search arrays rather than hardcode 0,0
+  const createReportAction = planDefinition.action?.find(
+    a => a.code[0].coding[0].code === 'create-report'
+  );
+
+  if (!createReportAction) {
+    throw new Error(`PlanDefinition ${planDefinition.id} has no action with code 'create-report'`);
+  }
+
+  const bundleOutput = createReportAction.output?.find(o => o.type === 'Bundle');
+
+  if (!bundleOutput?.profile?.[0]) {
+    throw new Error(
+      `PlanDefinition ${planDefinition.id} does not specify a profile for report bundle`
+    );
+    // option B is just use the default profile
+  }
+
+  return bundleOutput.profile[0];
+}
+
+function determineActionSequence(planDefinition) {
+  // TODO:
+  // note: 6.1.8
+  // The Backend Service App SHALL implement the ability
+  // to process action dependencies specified via
+  // the PlanDefinition.action.relatedAction elements.
+  // Typical implementations will build a graph or a tree structure
+  // that can be used to identify dependencies between actions.
+
+  // simple version to just get the integers 0 -> n
+  // https://stackoverflow.com/a/10050831
+  return [...Array(planDefinition.action.length).keys()];
+}
+
+function executeWorkflow(context) {
+  if (!context.currentActionSequenceStep) {
+    context.currentActionSequenceStep = 0;
+  }
+
+  const planDef = context.planDefinition;
+
+  if (!context.actionSequence) {
+    context.actionSequence = determineActionSequence(planDef);
+  }
+
+  db.upsert(COLLECTION, context, c => c.id === context.id);
+
+  const ig = findProfile(planDef);
+
+  while (context.currentActionSequenceStep < context.actionSequence.length) {
+    const currentActionId = context.actionSequence[context.currentActionSequenceStep];
+    const action = planDef.action[currentActionId];
+
+    // TODO - check arrays for
+    // system === "http://hl7.org/fhir/us/medmorph/CodeSystem/us-ph-plandefinition-actions"
+    const actionCode = action.code[0].coding[0].code;
+
+    const execute = getFunction(ig, actionCode);
+
+    // TODO: maintain context at each step for debugging?
+    // context = deepCopy(context);
+
+    // check the DB to get cancelToken
+    // TODO: is there a better way to do this?
+    const cancelToken = db.select(COLLECTION, c => c.id === context.id).cancelToken;
+    if (cancelToken) return;
+
+    execute(context);
+
+    // update the db after every completed step
```

Comment:
Is this to support timed events in the future? Everything else should run synchronously right, so the context passed is good enough. And then when a timed event runs it will pull the context from the db and call this method with it?

Reply:
Yeah that's the idea. I didn't think too deep into timed events, but I'm hoping this is a good enough start that we just have to plug a few things into, not necessarily rework the whole thing